cvxpy/lin_ops/canon_backend.py

```diff
@@ -1031,3 +1315,110 @@ def add_dicts(a: dict, b: dict) -> dict:
+                        param_slice_offset,
+                    ))
+        return TensorRepresentation.combine(tensor_representations)
+
+    def select_rows(self, rows: np.ndarray) -> None:
+
+        def func(x):
+            return x[:, rows, :]
+
+        self.apply_all(func)
+
+    def apply_all(self, func: Callable) -> None:
+        self.tensor = {var_id: {k: func(v)
+                                for k, v in parameter_repr.items()}
+                       for var_id, parameter_repr in self.tensor.items()}
+
+    def create_new_tensor_view(self, variable_ids: set[int], tensor: Any,
+                               is_parameter_free: bool) -> NumpyTensorView:
+        return NumpyTensorView(variable_ids, tensor, is_parameter_free, self.param_size_plus_one,
+                               self.id_to_col, self.param_to_size, self.param_to_col,
+                               self.var_length)
+
+    def accumulate_over_variables(self, func: Callable, is_param_free_function: bool) \
+            -> NumpyTensorView:
+        """
+        Apply 'func' to A and b.
+        If 'func' is a parameter free function, then we can apply it to all parameter slices
+        (including the slice that contains non-parameter constants).
+        If 'func' is not a parameter free function, we only need to consider the parameter slice
+        that contains the non-parameter constants, due to DPP rules.
+        """
+        for variable_id, tensor in self.tensor.items():
+            self.tensor[variable_id] = self.apply_to_parameters(func, tensor) if \
+                is_param_free_function else func(tensor[Constant.ID.value])
+
+        self.is_parameter_free = self.is_parameter_free and is_param_free_function
+        return self
+
+    @staticmethod
+    def apply_to_parameters(func: Callable,
+                            parameter_representation: dict[int, np.ndarray]) \
+            -> dict[int, np.ndarray]:
+        """
+        Apply 'func' to each slice of the parameter representation.
+        """
+        return {k: func(v) for k, v in parameter_representation.items()}
+
+    @staticmethod
+    def combine_potentially_none(a: Any | None, b: Any | None) -> Any | None:
+        if a is None and b is None:
+            return None
+        elif a is not None and b is None:
+            return a
+        elif a is None and b is not None:
+            return b
+        else:
+            return NumpyTensorView.add_dicts(a, b)
+
+    @staticmethod
+    def add_dicts(a: dict, b: dict) -> dict:
+        """
+        Addition for dict-based tensors.
+        """
+        res = {}
+        keys_a = set(a.keys())
+        keys_b = set(b.keys())
+        intersect = keys_a & keys_b
+        for key in intersect:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                res[key] = NumpyTensorView.add_dicts(a[key], b[key])
+            elif isinstance(a[key], list) and isinstance(b[key], list):
```

Comment:
We do not have lists in the numpy representation, right?

Reply:
should it be nd.array then?